src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolver.scala

```diff
@@ -0,0 +1,182 @@
+    val peerEventBus: ActorRef,
+    val etcPeerManager: ActorRef,
+    val blockchain: Blockchain,
+    val syncConfig: SyncConfig,
+    val appStateStorage: AppStateStorage,
+    implicit val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with PeerListSupport
+    with BlacklistSupport {
+
+  import syncConfig._
+
+  def handleCommonMessages: Receive = handlePeerListMessages orElse handleBlacklistMessages
+
+  override def receive: Receive = handleCommonMessages orElse { case StartBranchResolver =>
+    getBestPeer match {
+      case Some((peer, _)) => requestBlockHeadersGap(peer, blockchain.getBestBlockNumber())
+      case None =>
+        log.info(s"Still waiting for some peer, rescheduling StartBranchResolver")
+        scheduler.scheduleOnce(1.second, self, StartBranchResolver)
+    }
+  }
+
+  private def waitingBlockHeadersGap(bestBlockNumber: BigInt): Receive = handleCommonMessages orElse {
+    case ResponseReceived(peer, BlockHeaders(blockHeaders), timeTaken) => {
+      log.info("*** Received {} block headers in {} ms ***", blockHeaders.size, timeTaken)
+      getLastCorrectBlock(blockHeaders, bestBlockNumber) match {
+        case Some(lastValidBlockHeader) =>
+          finalizeBranchResolver(bestBlockNumber, lastValidBlockHeader)
+        case None => requestBlockHeader(peer, SearchState(0, bestBlockNumber))
+      }
+    }
+    case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+  }
+
+  private def binarySearchingLastValidBlock(searchState: SearchState): Receive = handleCommonMessages orElse {
+    case ResponseReceived(peer, BlockHeaders(blockHeaders), timeTaken) =>
+      log.info("*** Received {} block headers in {} ms ***", blockHeaders.size, timeTaken)
+      for {
+        parentBlockHeader <- blockHeaders.headOption
+        childBlockHeader <- blockchain.getBlockHeaderByNumber(parentBlockHeader.number - 1)
+      } yield validateBlockHeaders(parentBlockHeader, childBlockHeader, searchState, peer)
+
+    case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+  }
+
+  private def validateBlockHeaders(
+      parentBlockHeader: BlockHeader,
+      childBlockHeader: BlockHeader,
+      searchState: SearchState,
+      peer: Peer
+  ): Unit = {
+    if (validateBlockHeaders(parentBlockHeader, childBlockHeader)) {
+      if (childBlockHeader.number == searchState.minBlockNumber)
+        finalizeBranchResolver(blockchain.getBestBlockNumber(), childBlockHeader.number)
+      else requestBlockHeader(peer, searchState.copy(minBlockNumber = childBlockHeader.number + 1))
+    } else requestBlockHeader(peer, searchState.copy(maxBlockNumber = childBlockHeader.number))
+  }
+
+  private def finalizeBranchResolver(lastBlock: BigInt, lastValidBlock: BigInt): Unit = {
+    discardLastInvalidBlocks(lastBlock, lastValidBlock)
+    context become receive
+    log.info("branch resolution completed")
+    fastSync ! BranchResolvedSuccessful
```

Comment:
we should probably return 
- new master peer
- first common block with master peer